### PR TITLE
Adding ACL Policy Endpoints

### PR DESCRIPTION
**Pull request by armon**
This PR adds the ACL Policy endpoints and FSM methods to modify them. A follow up will add the HTTP endpoints.

### CODE REVIEW

File: nomad/structs/structs.go [hunk 1]
```diff
@@ -5236,9 +5238,59 @@ func IsRecoverable(e error) bool {
...
 
+// ACLPolicyListStub is used to for listing ACL policies
+type ACLPolicyListStub struct {
+	Name        string
+	Description string
 	CreateIndex uint64
 	ModifyIndex uint64
 }
+
+func (a *ACLPolicyListStub) FromPolicy(p *ACLPolicy) {
```
Comment (dadgar) [inline nomad/structs/structs.go]: Why don't you put this on ACLPolicy:

```
func (a *ACLPolicy) Stub() *ACLPolicyListStub {
  return ACLPolicyListStub{
    Name: a.Name,
    Description: a.Description,
    CreateIndex: a.CreateIndex,
    ModifyIndex: a.ModifyIndex,
  }
}
```

File: nomad/structs/structs.go [hunk 2]
```diff
@@ -5236,9 +5238,59 @@ func IsRecoverable(e error) bool {
 
 // ACLPolicy is used to represent an ACL policy
 type ACLPolicy struct {
-	Name  string // Unique name
-	Rules string // HCL or JSON format
+	Name        string // Unique name
```
Comment (dadgar) [inline nomad/structs/structs.go]: You may want to add a validate and ensure the name/description is length bounded

Comment (armon) [inline nomad/structs/structs.go]: Regex guards the name length, added a limit to the description as well.

File: nomad/acl_endpoint.go [hunk 3]
```diff
@@ -27,6 +39,19 @@ func (a *ACL) UpsertPolicies(args *structs.ACLPolicyUpsertRequest, reply *struct
 		return fmt.Errorf("must specify as least one policy")
 	}
 
+	// Validate each policy
+	for idx, policy := range args.Policies {
+		if !validPolicyName.MatchString(policy.Name) {
```
Comment (dadgar) [inline nomad/acl_endpoint.go]: @armon Would have preferred a `Validate() error` method on the policy and using multi error

**Pull request merged**